src/oncall/api/v0/upcoming_shifts.py

```diff
@@ -50,26 +52,34 @@ def on_get(req, resp, user_name):
     connection = db.connect()
     cursor = connection.cursor(db.DictCursor)
     role = req.get_param('role', None)
-    limit = req.get_param_as_int('limit', None)
+    limit = req.get_param_as_int('limit')
     query_end = ' ORDER BY `event`.`start` ASC'
-    query = '''SELECT %s, (SELECT COUNT(*) FROM `event` `counter`
-                           WHERE `counter`.`link_id` = `event`.`link_id`) AS num_events
+    query = '''SELECT %s
                FROM `event`
                JOIN `user` ON `user`.`id` = `event`.`user_id`
                JOIN `team` ON `team`.`id` = `event`.`team_id`
                JOIN `role` ON `role`.`id` = `event`.`role_id`
-               LEFT JOIN `event` `e2` ON `event`.link_id = `e2`.`link_id` AND `e2`.`start` < `event`.`start`
                WHERE `user`.`id` = (SELECT `id` FROM `user` WHERE `name` = %%s)
-                   AND `event`.`start` > UNIX_TIMESTAMP()
-                   AND `e2`.`start` IS NULL''' % all_columns
+                   AND `event`.`start` > UNIX_TIMESTAMP()''' % all_columns
 
     query_params = [user_name]
     if role:
         query_end = ' AND `role`.`name` = %s' + query_end
         query_params.append(role)
-    if limit:
-        query_end += ' LIMIT %s'
-        query_params.append(limit)
     cursor.execute(query + query_end, query_params)
     data = cursor.fetchall()
-    resp.body = json_dumps(data)
+    links = defaultdict(list)
+    formatted = []
+    for event in data:
+        if event['link_id'] is None:
+            formatted.append(event)
+        else:
+            links[event['link_id']].append(event)
+    for events in links.itervalues():
+        first_event = min(events, key = operator.itemgetter('start'))
+        first_event['num_events'] = len(events)
+        formatted.append(first_event)
+    formatted = sorted(formatted, key=operator.itemgetter('start'))
+    if limit is not None:
+        formatted = formatted[:limit]
```

Comment:
is it possible to do the limit in sql instead of using list slicing?

Reply:
Right, so the problem here is that I want to group together events by link_id and treat them as the "same event" (to avoid having your next 3 upcoming shift all be part of the same 12-hour event). Doing this aggregation in SQL is really awkward and requires a self join to be guaranteed correct, which makes a massive query that's slow and terrible. Technically we don't need the self join, but then the query's correctness relies on data characteristics that aren't enforced in the DB. I did some performance testing, and processing in Python is a lot better unless we have a huge amount of events for that user in the future, which is unlikely.